source/app/controllers/urls_controller.rb

```diff
@@ -1,2 +1,47 @@
 class UrlsController < ApplicationController
+  # GET /urls
+  # GET /urls.json
+  def index
+    redirect_to login_path and return if !logged_in?
+    @urls = Url.all
+  end
+
+  # GET /urls/1
+  # GET /urls/1.json
+  def show
+    url = Url.find_by(short_url: params[:id])
```

Comment:
`find_by` returns `nil` when a record isn't found. Since the lines below this depend on `url` being non-nil, switch this to `find_by!`. `find_by!` raises `ActiveRecord::RecordNotFound` instead. Rails automatically responds with a 404 when `ActiveRecord::RecordNotFound` is raised.